[CA-1359] A11y: Skip Navigation Links

andrewmarcus opened the pull request:
This functionality was extracted from #2501, where it was already reviewed and approved. I'm splitting that PR up into smaller components to try to figure out what's causing the integration tests to fail.

### Skip Navigation Link

Most websites provide a link, first in the tab order on the page, allowing screen reader users to skip repetitive content at the top of the page. This is a standard design practice which users expect to be present. In this case there isn't a ton of content in the header, but it still seems useful to be able to skip it.

Because of the way React passes refs down to child components in order to be able to target elements for focusing, I decided it was important to encapsulate all of the skip functionality in as few places as possible, to avoid having to update every page to accept an addition forwarded ref. I was able to set it up entirely within the TopBar, so the skip link is first, and the skip target is last, and appears visually as focus on the entire page. The next tab stop is the first element below the header.

The one potential problem with this approach is that the header isn't entirely identical on every page. The `<h1>` is in it on most pages except the landing page, and it sometimes also contains contextual info such as breadcrumbs, a search bar or info about the billing project. Consequently a user could skip things that are relevant to the page without knowing it. However, these elements still appear visually and semantically within the header, so it seems reasonable. It's a pattern users of assistive technologies expect, and on some pages the header is quite extensive.

andrewmarcus committed:
Added a Skip Navigation link and target to the TopBar. This was extracted from the a11y/awm-workspaces-list branch for improved encapsulation and testability

andrewmarcus committed:
Merge remote-tracking branch 'origin/dev' into a11y/awm-skip-navlnk

s-rubenstein reviewed (approved):
This seems reasonable to me.

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-skip-navlnk

dmohs commented on src/components/skipNavLink.js:
```diff
@@ -0,0 +1,42 @@
+    className: 'reveal-on-focus',
+    style: {
+      position: 'absolute',
+      top: 0,
+      left: 0,
```
Nit: One line?

dmohs commented on src/components/skipNavLink.js:
```diff
@@ -0,0 +1,42 @@
+      left: 0,
+      background: 'white',
+      color: colors.accent(),
+      border: '1px solid black',
+      borderColor: colors.accent(),
```
Just curious why you did this instead of:
```suggestion
      border: '1px solid',
      borderColor: colors.accent(),
```

andrewmarcus replied:
Habit, I think. ;-)  Thanks for pointing it out

dmohs commented on src/components/skipNavLink.js:
```diff
@@ -0,0 +1,42 @@
+      color: colors.accent(),
+      border: '1px solid black',
+      borderColor: colors.accent(),
+      padding: '1em',
```
Stick with `rem` here? We try to discourage use of `em` in the codebase since people don't always expect it to be different.

dmohs commented on src/components/skipNavLink.js:
```diff
@@ -0,0 +1,42 @@
+      event.stopPropagation()
+      ref.current?.focus()
+    },
+    ...props
```
`_.merge` here to allow merging of styles?

andrewmarcus replied:
I added `_.merge` to the `SkipNavLink` component. For some reason it wasn't working properly in the `SkipNavTarget` component, probably because of the `ref`, so I left that one as-is.

dmohs commented on src/components/skipNavLink.js:
```diff
@@ -0,0 +1,42 @@
+      borderColor: colors.accent(),
+      padding: '1em',
+      boxShadow: 'rgba(0, 0, 0, 0.5) 0px 0px 4px 0px',
+      zIndex: 9998
```
We trust your judgment that this is the best way to do this, but would you also please add a line comment?

andrewmarcus replied:
Sure, make sense. I mostly just need to ensure that the button and outline both appear on top of anything else defined after them in the DOM

dmohs commented on src/style.css:
```diff
@@ -151,3 +151,22 @@ a {
+  bottom: 0;
+  left: 0;
+  right: 0;
+  z-index: 9999;
```
Reciprocal line comment here?

andrewmarcus committed:
Update src/components/skipNavLink.js

Removed unnecessary extra border color

Co-authored-by: David Mohs <dmohs@broadinstitute.org>

andrewmarcus committed:
PR comments:
* Use _.merge() for SkipNavLink component properties
* Switched to using rem for measurements

Also, I dropped the top of the SkipNavTarget focus ring below the banner, which makes it even clearer what's going on
* Moved SkipNavTarget styling directly into component, for clarity and consistency

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-skip-navlnk

# Conflicts:
#	src/style.css

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-skip-navlnk

# Conflicts:
#	src/components/TopBar.js
